provider/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt

```diff
@@ -204,44 +205,60 @@ fun kotlinCoreEnvironmentFor(configuration: CompilerConfiguration, rootDisposabl
 
 
 internal
-fun messageCollectorFor(log: Logger, pathTranslation: (String) -> String = { it }): MessageCollector =
+fun messageCollectorFor(log: Logger, pathTranslation: (String) -> String = { it }): LoggingMessageCollector =
+    LoggingMessageCollector(log, pathTranslation)
 
-    object : MessageCollector {
 
-        var errors = 0
+internal
+data class ScriptCompilationError(val message: String, val lineNumber: Int?)
 
-        override fun hasErrors() = errors > 0
 
-        override fun clear() { errors = 0 }
+internal
+data class ScriptCompilationException(val errors: List<ScriptCompilationError>) : RuntimeException() {
+    override val message: String?
+        get() = errors.first().message
```

Comment:
Same question as above:
https://github.com/gradle/kotlin-dsl/pull/688/files#r163057841

Reply:
It is not guaranteed. Fixed